### ISSUE TITLE
Unwrapping an Option but not using the value is probably unintended

### ISSUE THREAD

**Issue by thombles** [rust-lang/rust-clippy#13593]
### What it does

Suppose you call a method that returns an `Option<String>`. Calling `unwrap()` on the result but failing to use the value is suspicious as a correctness issue. Unlike a `Result<(), _>` where unwrapping is a common way to to assert `.is_ok()`, unwrapping an `Option` and discarding the value is a little odd.

### Advantage

* This catches refactoring issues where an existing `unwrap()` used to apply to a `Result` (presumed infallible) but now applies to an `Option` (because we have now addressed the infallibility). It still compiles but has a completely different meaning.
    * (Real world example - tokio's `watch::Sender`'s [`send()`](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send) vs [`send_replace()`](https://docs.rs/tokio/latest/tokio/sync/watch/struct.Sender.html#method.send_replace) where `T` is an `Option<_>`)

### Drawbacks

It is possible this is used inside tests to induce a panic. Personally I would use `assert!(value.is_some())` but it's possible there's a lot of existing code which simply unwraps.

### Example

```rust
fn foo() -> Option<String> {
    Some("bar".to_owned())
}

fn main() {
    foo().unwrap(); // unwrapped value is unused
}
```

Could be written as:

```rust
fn foo() -> Option<String> {
    Some("bar".to_owned())
}

fn main() {
    foo();
}
```

**Comment by samueltardieu** [comment]
Maybe this could be added as a `style`-category lint, as it might depend on the coding style chosen for a given project. An `Option` used as a statement would trigger the lint.

**Comment by GnomedDev** [comment]
If you feel like `Option::unwrap` should lint like this, this should be an issue on the rustc repo to just add `#[must_use]` to `Option::unwrap`, but I doubt that would be accepted and I don't think this should be a part of clippy.

**Comment by thombles** [comment]
> An `Option` used as a statement would trigger the lint.

Hopefully not! Only unwrapping it.

> If you feel like `Option::unwrap` should lint like this, this should be an issue on the rustc repo ... I don't think this should be a part of clippy.

Fair enough!
